File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
'''
1.아이디어
    - 상대진영인 maps[len(maps)][len(mpas[0])]에 도달하는 최단경로 문제(BFS/Queue)
    1) maps의 상하좌우를 확인하여, 1인 곳으로 이동 + 지나온 곳은 0으로 변경
    2) 더이상 갈 곳이 없다면 -1 반환
    3) queue에다가 다음 길을 넣고, 꺼내서 그 다음 경로로 쭉 계산
    4) 상대 진영에 도착했을 때의 cnt 값 저장
'''
from collections import deque


def solution(maps):
    n, m = len(maps), len(maps[0])
    visited = [[False]*m for _ in range(n)]
    queue = deque([(0, 0, 1)])  # (x, y, distance)
    visited[0][0] = True

    dx = [-1, 1, 0, 0]  # 상하좌우
    dy = [0, 0, -1, 1]

    while queue:
        x, y, dist = queue.popleft()

        if x == n-1 and y == m-1:
            return dist  # 도착

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]

            if 0 <= nx < n and 0 <= ny < m:
                if maps[nx][ny] == 1 and not visited[nx][ny]:
                    visited[nx][ny] = True
                    queue.append((nx, ny, dist + 1))

    return -1  # 도달 불가

        
        
    
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
/*
Q. 출발지점(1,1)에서 도착지점(N,M)까지 최단거리로 가는 방법
A. BFS를 이용해서 도착지점에 가는 최단거리 구하기
*/
import java.util.*;

class Solution {
    
    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};
    
    public int solution(int[][] maps) {
        
        // 변수 초기화
        int n = maps.length;
        int m = maps[0].length;
        
        // 도착지점에 도달할 수 있으면, 최단거리 찾기
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        
        // 시작점(0,0)에서 출발
        q.add(new int[] {0,0,1}); 
        visited[0][0] = true;
        
        while(!q.isEmpty()){
            int[] cur = q.poll();
            
            // 도착점이라면 종료
            if(cur[0] == n-1 && cur[1] == m-1){
                return cur[2];
            }
            
            // 시작점에서 갈 수 있는 길 확인하기
            for(int d=0; d<4; d++){
                int ny = cur[0] + dy[d];
                int nx = cur[1] + dx[d];
                if(ny<0 || ny>=n || nx<0 || nx>=m) continue; 
                if(maps[ny][nx] == 0) continue;
                
                // 벽이 아닌 길이라면, 이동
                if(!visited[ny][nx]){
                    q.add(new int[] {ny, nx, cur[2]+1});
                    visited[ny][nx] = true;
                }
            }
        }
        
        return -1;
        
    }
}